File: mobile/ios/OpenU60/Features/RouterSettings/Firewall/FirewallSettingsViewModel.swift
import SwiftUI

@Observable
@MainActor
final class FirewallSettingsViewModel {
    var config: FirewallConfig = .empty
    var portForwardRules: [PortForwardRule] = []
    var filterRules: [FilterRule] = []
    var upnpEnabled: Bool = false
    var isLoading: Bool = false
    var message: String?
    var messageIsError: Bool = false
    var showAddPortForward: Bool = false

    // DMZ edit fields
    var editDmzEnabled: Bool = false
    var editDmzIP: String = ""

    private let client: AgentClient
    private let authManager: AuthManager

    init(client: AgentClient, authManager: AuthManager) {
        self.client = client
        self.authManager = authManager
    }

    func refresh() async {
        isLoading = true
        message = nil

        do {
            let data = try await client.getJSON("/api/router/firewall")
            config = FirewallParser.parseConfig(data)
            editDmzEnabled = config.dmzEnabled
            editDmzIP = config.dmzHost

            async let pfResult = fetchPortForwardRules()
            async let filterResult = fetchFilterRules()
            async let upnpResult = fetchUPnP()

            portForwardRules = await pfResult
            filterRules = await filterResult
            upnpEnabled = await upnpResult
        } catch {
            showMessage("Failed to load firewall: \(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    private func fetchPortForwardRules() async -> [PortForwardRule] {
        do {
            let data = try await client.getJSON("/api/router/firewall/port-forward")
            return FirewallParser.parsePortForwardRules(data)
        } catch {
            return []
        }
    }

    private func fetchFilterRules() async -> [FilterRule] {
        do {
            let data = try await client.getJSON("/api/router/firewall/filter-rules")
            return FirewallParser.parseFilterRules(data)
        } catch {
            return []
        }
    }

    private func fetchUPnP() async -> Bool {
        do {
            let data = try await client.getJSON("/api/router/firewall/upnp")
            if let str = data["upnp_switch"] as? String {
                return str == "1"
            }
            return false
        } catch {
            return false
        }
    }

    func toggleFirewall(enabled: Bool) async {
        isLoading = true
        do {
            let _ = try await client.putJSON("/api/router/firewall/switch", body: ["firewall_switch": enabled ? "1" : "0"])
            showMessage("Firewall \(enabled ? "enabled" : "disabled")", isError: false)
            config.enabled = enabled
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }
        isLoading = false
    }

    func setLevel(_ level: String) async {
        isLoading = true
        do {
            let _ = try await client.putJSON("/api/router/firewall/level", body: ["firewall_level": level])
            showMessage("Firewall level set to \(level)", isError: false)
            config.level = level
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }
        isLoading = false
    }

    func toggleNAT(enabled: Bool) async {
        isLoading = true
        do {
            let _ = try await client.putJSON("/api/router/firewall/nat", body: ["nat_switch": enabled ? "1" : "0"])
            showMessage("NAT \(enabled ? "enabled" : "disabled")", isError: false)
            config.nat = enabled
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }
        isLoading = false
    }

    func toggleUPnP(enabled: Bool) async {
        isLoading = true
        do {
            let _ = try await client.putJSON("/api/router/firewall/upnp", body: ["upnp_switch": enabled ? "1" : "0"])
            showMessage("UPnP \(enabled ? "enabled" : "disabled")", isError: false)
            upnpEnabled = enabled
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }
        isLoading = false
    }

    func applyDMZ() async {
        isLoading = true
        do {
            let _ = try await client.putJSON("/api/router/firewall/dmz", body: [
                "dmz_enabled": editDmzEnabled ? "1" : "0",
                "dmz_ip": editDmzIP
            ])
            showMessage("DMZ settings updated", isError: false)
            config.dmzEnabled = editDmzEnabled
            config.dmzHost = editDmzIP
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }
        isLoading = false
    }

    func togglePortForward(enabled: Bool) async {
        isLoading = true
        do {
            let _ = try await client.putJSON("/api/router/firewall/port-forward/switch", body: ["port_forward_switch": enabled ? "1" : "0"])
            showMessage("Port forwarding \(enabled ? "enabled" : "disabled")", isError: false)
            config.portForwardEnabled = enabled
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }
        isLoading = false
    }

    func addPortForward(name: String, protocol_: String, wanPort: String, lanIP: String, lanPort: String) async {
        isLoading = true
        do {
            let _ = try await client.postJSON("/api/router/firewall/port-forward", body: [
                "action": "add",
                "name": name,
                "protocol": protocol_,
                "wan_port": wanPort,
                "lan_ip": lanIP,
                "lan_port": lanPort,
                "enabled": "1"
            ])
            showAddPortForward = false
            showMessage("Port forward rule added", isError: false)
            portForwardRules.append(PortForwardRule(id: UUID().uuidString, name: name, protocol_: protocol_, wanPort: wanPort, lanIP: lanIP, lanPort: lanPort, enabled: true))
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }
        isLoading = false
    }

    func deletePortForward(_ rule: PortForwardRule) async {
        isLoading = true
        do {
            let _ = try await client.postJSON("/api/router/firewall/port-forward", body: [
                "action": "delete",
                "id": rule.id
            ])
            showMessage("Port forward rule deleted", isError: false)
            portForwardRules.removeAll { $0.id == rule.id }
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }
        isLoading = false
    }

    private func showMessage(_ text: String, isError: Bool) {
        message = text
        messageIsError = isError
    }
}

File: mobile/android/OpenU60/app/src/main/java/com/openu60/feature/router/firewall/FirewallSettingsViewModel.kt
package com.openu60.feature.router.firewall

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.openu60.core.model.FirewallConfig
import com.openu60.core.model.FirewallParser
import com.openu60.core.model.PortForwardRule
import com.openu60.core.network.AgentClient
import com.openu60.core.network.AgentError
import com.openu60.core.network.AuthManager
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class FirewallSettingsState(
    val config: FirewallConfig = FirewallConfig.empty,
    val portForwardRules: List<PortForwardRule> = emptyList(),
    val isLoading: Boolean = false,
    val message: String? = null,
    val messageIsError: Boolean = false,
    val showPortForwardForm: Boolean = false,
)

@HiltViewModel
class FirewallSettingsViewModel @Inject constructor(
    private val agentClient: AgentClient,
    private val authManager: AuthManager,
) : ViewModel() {

    private val _state = MutableStateFlow(FirewallSettingsState())
    val state: StateFlow<FirewallSettingsState> = _state.asStateFlow()

    fun refresh() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                val configData = agentClient.getJSON("/api/firewall/config")
                val config = FirewallParser.parseConfig(configData)

                val pfData = agentClient.getJSON("/api/firewall/port-forward")
                val rules = FirewallParser.parsePortForwardRules(pfData)

                _state.value = _state.value.copy(config = config, portForwardRules = rules, isLoading = false)
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) refresh() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun toggleFirewall(enabled: Boolean) = updateConfig(mapOf("firewall_switch" to if (enabled) "1" else "0"))
    fun toggleNAT(enabled: Boolean) = updateConfig(mapOf("nat_switch" to if (enabled) "1" else "0"))
    fun toggleUPnP(enabled: Boolean) = updateConfig(mapOf("upnp_switch" to if (enabled) "1" else "0"))
    fun togglePortForward(enabled: Boolean) = updateConfig(mapOf("port_forward_switch" to if (enabled) "1" else "0"))

    fun setFirewallLevel(level: String) = updateConfig(mapOf("firewall_level" to level))

    fun setDMZ(enabled: Boolean, host: String) = updateConfig(mapOf(
        "dmz_enabled" to if (enabled) "1" else "0",
        "dmz_ip" to host,
    ))

    fun addPortForwardRule(name: String, protocol: String, wanPort: String, lanIP: String, lanPort: String, enabled: Boolean) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.postJSON("/api/firewall/port-forward", mapOf(
                    "name" to name,
                    "protocol" to protocol,
                    "wan_port" to wanPort,
                    "lan_ip" to lanIP,
                    "lan_port" to lanPort,
                    "enabled" to if (enabled) "1" else "0",
                ))
                _state.value = _state.value.copy(
                    showPortForwardForm = false,
                    message = "Rule added",
                    messageIsError = false,
                )
                refresh()
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) addPortForwardRule(name, protocol, wanPort, lanIP, lanPort, enabled)
                else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun deletePortForwardRule(id: String) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.deleteJSON("/api/firewall/port-forward", mapOf("id" to id))
                _state.value = _state.value.copy(message = "Rule deleted", messageIsError = false)
                refresh()
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) deletePortForwardRule(id) else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun showAddForm() {
        _state.value = _state.value.copy(showPortForwardForm = true)
    }

    fun hideAddForm() {
        _state.value = _state.value.copy(showPortForwardForm = false)
    }

    private fun updateConfig(params: Map<String, Any?>) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.putJSON("/api/firewall/config", params)
                _state.value = _state.value.copy(message = "Settings updated", messageIsError = false)
                refresh()
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) updateConfig(params) else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    private fun setError(msg: String?) {
        _state.value = _state.value.copy(isLoading = false, message = msg ?: "Unknown error", messageIsError = true)
    }
}
